File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/discovery/repository/DiscoveryRepository.java
package com.ixuea.courses.mymusic.component.discovery.repository;

import com.ixuea.courses.mymusic.component.ad.model.Ad;
import com.ixuea.courses.mymusic.component.discovery.model.DiscoveryPage;
import com.ixuea.courses.mymusic.component.discovery.model.ui.BannerData;
import com.ixuea.courses.mymusic.component.discovery.model.ui.BaseSort;
import com.ixuea.courses.mymusic.component.discovery.model.ui.ButtonData;
import com.ixuea.courses.mymusic.component.discovery.model.ui.FooterData;
import com.ixuea.courses.mymusic.component.discovery.model.ui.SheetData;
import com.ixuea.courses.mymusic.component.discovery.model.ui.SongData;
import com.ixuea.courses.mymusic.component.sheet.model.Sheet;
import com.ixuea.courses.mymusic.component.song.model.Song;
import com.ixuea.courses.mymusic.model.response.ListResponse;
import com.ixuea.courses.mymusic.model.ui.BaseMultiItemEntity;
import com.ixuea.courses.mymusic.repository.DefaultRepository;
import com.ixuea.courses.mymusic.util.Constant;
import com.ixuea.courses.mymusic.util.PreferenceUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import io.reactivex.rxjava3.core.Observable;

/**
 * Builds typed immutable sections for the discovery page.
 */
public class DiscoveryRepository {
    private static DiscoveryRepository instance;
    private final DefaultRepository repository;

    private DiscoveryRepository(DefaultRepository repository) {
        this.repository = repository;
    }

    public synchronized static DiscoveryRepository getInstance() {
        if (instance == null) {
            instance = new DiscoveryRepository(DefaultRepository.getInstance());
        }
        return instance;
    }

    public Observable<DiscoveryPage> homeSections(PreferenceUtil sp) {
        return Observable.zip(
                repository.bannerAd(),
                repository.sheets(Constant.SIZE12),
                repository.songs(),
                (ads, sheets, songs) -> new DiscoveryPage(buildSections(sp, ads, sheets, songs))
        );
    }

    private List<BaseMultiItemEntity> buildSections(
            PreferenceUtil sp,
            ListResponse<Ad> ads,
            ListResponse<Sheet> sheets,
            ListResponse<Song> songs
    ) {
        ArrayList<BaseMultiItemEntity> sections = new ArrayList<>();
        sections.add(new BannerData(ads.getData().getData(), sp.getSort(Constant.STYLE_BANNER)));
        sections.add(new ButtonData(sp.getSort(Constant.STYLE_BUTTON)));
        sections.add(new SheetData(sheets.getData().getData(), sp.getSort(Constant.STYLE_SHEET)));
        sections.add(new SongData(songs.getData().getData(), sp.getSort(Constant.STYLE_SONG)));
        sections.add(new FooterData());
        Collections.sort(sections, (o1, o2) -> ((BaseSort) o1).compareTo((BaseSort) o2));
        return Collections.unmodifiableList(sections);
    }
}

File: code/video/MyCloudMusicAndroidJava/app/src/main/java/com/ixuea/courses/mymusic/component/discovery/repository/DiscoveryRepository.kt
package com.ixuea.courses.mymusic.component.discovery.repository

import com.ixuea.courses.mymusic.component.ad.model.Ad
import com.ixuea.courses.mymusic.component.discovery.model.DiscoveryPage
import com.ixuea.courses.mymusic.component.discovery.model.ui.BannerData
import com.ixuea.courses.mymusic.component.discovery.model.ui.BaseSort
import com.ixuea.courses.mymusic.component.discovery.model.ui.ButtonData
import com.ixuea.courses.mymusic.component.discovery.model.ui.FooterData
import com.ixuea.courses.mymusic.component.discovery.model.ui.SheetData
import com.ixuea.courses.mymusic.component.discovery.model.ui.SongData
import com.ixuea.courses.mymusic.component.sheet.model.Sheet
import com.ixuea.courses.mymusic.component.song.model.Song
import com.ixuea.courses.mymusic.model.response.ListResponse
import com.ixuea.courses.mymusic.model.ui.BaseMultiItemEntity
import com.ixuea.courses.mymusic.repository.DefaultRepository
import com.ixuea.courses.mymusic.util.Constant
import com.ixuea.courses.mymusic.util.PreferenceUtil
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope

/**
 * Builds typed immutable sections for the discovery page.
 */
class DiscoveryRepository private constructor(
    private val repository: DefaultRepository,
) {
    suspend fun homeSections(sp: PreferenceUtil): DiscoveryPage = coroutineScope {
        val ads = async { repository.bannerAd() }
        val sheets = async { repository.sheets(Constant.SIZE12) }
        val songs = async { repository.songs() }

        DiscoveryPage(
            buildSections(
                sp,
                ads.await(),
                sheets.await(),
                songs.await(),
            )
        )
    }

    private fun buildSections(
        sp: PreferenceUtil,
        ads: ListResponse<Ad>,
        sheets: ListResponse<Sheet>,
        songs: ListResponse<Song>,
    ): List<BaseMultiItemEntity> {
        val sections = mutableListOf<BaseMultiItemEntity>()
        sections += BannerData(ads.data?.data.orEmpty().toMutableList(), sp.getSort(Constant.STYLE_BANNER))
        sections += ButtonData(sp.getSort(Constant.STYLE_BUTTON))
        sections += SheetData(sheets.data?.data.orEmpty().toMutableList(), sp.getSort(Constant.STYLE_SHEET))
        sections += SongData(songs.data?.data.orEmpty().toMutableList(), sp.getSort(Constant.STYLE_SONG))
        sections += FooterData()
        sections.sortWith { first, second ->
            (first as BaseSort).compareTo(second as BaseSort)
        }
        return sections.toList()
    }

    companion object {
        @Volatile
        private var instance: DiscoveryRepository? = null

        @JvmStatic
        fun getInstance(): DiscoveryRepository {
            return instance ?: synchronized(this) {
                instance ?: DiscoveryRepository(DefaultRepository.getInstance()).also {
                    instance = it
                }
            }
        }
    }
}
